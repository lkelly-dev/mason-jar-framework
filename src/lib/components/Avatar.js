import React, { useContext } from 'react';
import { cx } from 'emotion';
import MODIFIERS from './Modifiers';
import { debugAttrs, debugMode } from '../utils/debugging';

const Avatar = props => {
  const { context, children, size, status, className, ...rest } = props;
  const c = (context) ? useContext(context) : null;
  const debug = debugMode(c)
  const avatarSize = MODIFIERS.SIZE[size] ? `avatar-${size}` : null;
  const avatarStatus = MODIFIERS.STATUS[status] ? `avatar-${status}` : null;
  const avatarClass = cx('avatar', avatarSize, avatarStatus, className);
  return (
    <div className={avatarClass}
      {...(debug) ? debugAttrs(props) : null}
      {...rest}>
      {children}
    </div>
  );
};

const AvatarImage = props => {
  const { context, shape, alt, className, ...rest } = props;
  const c = (context) ? useContext(context) : null;
  const debug = debugMode(c)
  const avatarImgShape = MODIFIERS.SHAPE[shape] || null;
  const avatarImgClass = cx('avatar-img', avatarImgShape, className);
  return <img className={avatarImgClass} alt={alt || ''}
              {...(debug) ? debugAttrs(props) : null}
              {...rest} />;
};

const AvatarTitle = props => {
  const { context, children, shape, className, ...rest } = props;
  const c = (context) ? useContext(context) : null;
  const debug = debugMode(c)
  const avatarTitleShape = MODIFIERS.SHAPE[shape] || null;
  const avatarTitleClass = cx('avatar-title', avatarTitleShape, className);
  return (
    <span className={avatarTitleClass}
      {...(debug) ? debugAttrs(props) : null}
      {...rest}>
      {children}
    </span>
  );
};

Avatar.Image = AvatarImage;
Avatar.Title = AvatarTitle;

export { Avatar, AvatarImage, AvatarTitle };
