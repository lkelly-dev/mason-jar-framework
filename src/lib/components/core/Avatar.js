import React from 'react';
import { cx } from 'emotion';
import MODIFIERS from './Modifiers';

const Avatar = props => {
  const { children, size, status, className, ...rest } = props;
  const avatarSize = MODIFIERS.SIZE[size] ? `avatar-${size}` : null;
  const avatarStatus = MODIFIERS.STATUS[status] ? `avatar-${status}` : null;
  const avatarClass = cx('avatar', avatarSize, avatarStatus, className);
  return (
    <div className={avatarClass} {...rest}>
      {children}
    </div>
  );
};

const AvatarImage = props => {
  const { shape, alt, className, ...rest } = props;
  const avatarImgShape = MODIFIERS.SHAPE[shape] || null;
  const avatarImgClass = cx('avatar-img', avatarImgShape, className);
  return <img className={avatarImgClass} alt={alt || ''} {...rest} />;
};

const AvatarTitle = props => {
  const { children, shape, className, ...rest } = props;
  const avatarTitleShape = MODIFIERS.SHAPE[shape] || null;
  const avatarTitleClass = cx('avatar-title', avatarTitleShape, className);
  return (
    <span className={avatarTitleClass} {...rest}>
      {children}
    </span>
  );
};

Avatar.Image = AvatarImage;
Avatar.Title = AvatarTitle;

export { Avatar, AvatarImage, AvatarTitle };
