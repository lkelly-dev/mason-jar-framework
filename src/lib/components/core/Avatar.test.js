import React from 'react';
import { shallow } from 'enzyme';
import { Avatar, AvatarImage, AvatarTitle } from "./Avatar";

describe('Avatar', () => {

  it('matches snapshot', () => {
    const component = shallow(<Avatar />);
    expect(component).toMatchSnapshot();
  });

  it('contains children', () => {
    const component = shallow(<Avatar>A</Avatar>);
    expect(component.props().children).toBe("A");
  });

  it('has expected classnames', () => {
    const child = <div>child</div>,
          size = 'sm',
          status = 'offline',
          cn = 'nifty';
    const componentBasic = shallow(<Avatar>{ child }</Avatar>);
    const componentWithOptions = shallow(
      <Avatar
        size={ size }
        status= { status }
        className={ cn }
        whatever="dude">
        { child }
      </Avatar>
    );
    const classNameArray = componentWithOptions.prop('className').split(' ');
    expect(componentBasic.props().className).toBe('avatar');
    expect(componentBasic.contains(child)).toEqual(true);
    expect(classNameArray).toContain('avatar');
    expect(classNameArray).toContain(`avatar-${size}`);
    expect(classNameArray).toContain(`avatar-${status}`);
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(4)
    expect(componentWithOptions.prop('whatever')).toBe('dude');
  });

  it('doesnt use invalid prop values', () => {
    const size = 'wicked-big',
          status = 'single';
    const component = shallow(<Avatar size={ size } status={ status }/>);
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).not.toContain(`avatar-${size}`);
    expect(classNameArray).not.toContain(`avatar-${status}`);
    expect(classNameArray).toHaveLength(1)
  });
});

describe('AvatarImage', () => {

  it('matches snapshot', () => {
    const component = shallow(<AvatarImage />);
    expect(component).toMatchSnapshot();
  });

  it('has correct props', () => {
    const component = shallow(<AvatarImage shape={ shape } alt="User Avatar" className={ cn } whatever="dude" />);
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('avatar-img');
    expect(classNameArray).toContain(shape);
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(3);
    expect(component.prop('alt')).toBe("User Avatar");
    expect(component.prop('whatever')).toBe('dude');
  });

  it('doesnt use invalid prop values', () => {
    const component = shallow(<AvatarImage shape={ badShape }/>);
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).not.toContain(badShape);
    expect(classNameArray).toHaveLength(1)
  });
});

describe('AvatarTitle', () => {

  it('matches snapshot', () => {
    const component = shallow(<AvatarTitle />);
    expect(component).toMatchSnapshot();
  });

  it('has correct props', () => {
    const child = <div>child</div>;
    const component = shallow(
      <AvatarTitle
        shape={ shape }
        className={ cn }
        whatever="dude">
        { child }
      </AvatarTitle>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('avatar-title');
    expect(classNameArray).toContain(shape);
    expect(classNameArray).toContain(cn);
    expect(classNameArray).toHaveLength(3);
    expect(component.prop('whatever')).toBe('dude');
    expect(component.contains(child)).toEqual(true)
  });

  it('doesnt use invalid prop values', () => {
    const component = shallow(<AvatarTitle shape={ badShape }/>);
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).not.toContain(badShape);
    expect(classNameArray).toHaveLength(1)
  });
});
