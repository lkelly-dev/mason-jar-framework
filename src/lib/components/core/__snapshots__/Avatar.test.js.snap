// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar matches snapshot 1`] = `
<div
  className="avatar"
/>
`;

exports[`AvatarImage matches snapshot 1`] = `
<img
  alt=""
  className="avatar-img"
/>
`;

exports[`AvatarTitle matches snapshot 1`] = `
<span
  className="avatar-title"
/>
`;
